src/common/src/array/struct_array.rs

```diff
@@ -434,6 +441,65 @@ impl ToText for StructRef<'_> {
     }
 }
 
+/// Double quote a string if it contains any special characters.
+fn quote_if_need(input: &str, writer: &mut impl Write) -> std::fmt::Result {
+    if !input.is_empty() // non-empty
+        && !input.contains([
+            '"', '\\', '(', ')', ',',
+            // PostgreSQL `array_isspace` includes '\x0B' but rust
+            // [`char::is_ascii_whitespace`] does not.
+            ' ', '\t', '\n', '\r', '\x0B', '\x0C',
+        ])
+    {
+        return writer.write_str(input);
+    }
+
+    writer.write_char('"')?;
+
+    for ch in input.chars() {
+        match ch {
+            '"' => writer.write_str("\"\"")?,
+            '\\' => writer.write_str("\\\\")?,
+            _ => writer.write_char(ch)?,
+        }
+    }
+
+    writer.write_char('"')
+}
+
+/// Remove double quotes from a string.
+/// This is the reverse of [`quote_if_need`].
+#[allow(dead_code)]
+fn unquote_if_need(mut input: &str) -> Cow<'_, str> {
+    if input.starts_with('"') && input.ends_with('"') {
+        input = &input[1..input.len() - 1];
+    }
```

Comment:
The rules on inner quotes are different in PostgreSQL with or without outer quotes:
```
test=# select '(1,"a""b")'::z;
     z      
------------
 (1,"a""b")
(1 row)

test=# select '(1,a""b)'::z;
   z    
--------
 (1,ab)
(1 row)
```
It even allows entering and exiting quotes multiple times (unlike in array):
```
test=# select '(1,"a"b"c"d"e")'::z;
     z     
-----------
 (1,abcde)
(1 row)

test=# select '(1,a"b"c"d"e)'::z;
     z     
-----------
 (1,abcde)
(1 row)
```

The precise semantic of this function would depend on how it is invoked by the parser, which is not implemented yet. We have several options here:
* Remove this function until we fully implement varchar -> struct
* Define it to only reverse `quote_if_need` and used by unit tests. Acknowledge in comment it is different from `unescape` needed by `record_in` (aka varchar -> struct)

Reply:
It sounds too complicated. 🥵 I'd like to remove this function for now.